data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,916 @@
+            .for_each(|entry| {
+                fees.push(entry);
+            });
+
+        Self { priorities: fees }
+    }
+
+    /// Get the entry at a certain position, if an item at that position exists, or None otherwise.
+    #[inline]
+    pub fn get(&self, index: usize) -> Option<&Priorities> {
+        if index >= self.priorities.capacity() {
+            None
+        } else {
+            self.priorities.iter().nth(index)
+        }
+    }
+
+    /// Push a new `Priorities` entry into the engine.
+    #[inline]
+    pub fn push_priorities(&mut self, priorities: Priorities) {
+        log::trace!("Pushing new transaction priorities entry: {:?}", priorities);
+        self.priorities.push(priorities);
+        log::trace!(
+            "The priority engine has received new data. The priority estimate is now:\n{}",
+            self.estimate_priority().unwrap_or_default()
+        );
+    }
+
+    /// Create a new engine of a certain queue capacity.
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            priorities: CircularQueue::with_capacity(capacity),
+        }
+    }
+}
+
+/// Different errors that the `PriorityEngine` can produce.
+#[derive(Debug, Eq, Fail, PartialEq)]
+pub enum PriorityError {
+    /// The number of sampled epochs in the engine is not enough for providing a reliable estimate.
+    #[fail(
+        display = "The node has only sampled priority from {} blocks but at least {} are needed to provide a reliable priority estimate. Please retry after {} minutes.",
+        _0, _1, _2
+    )]
+    NotEnoughSampledEpochs(usize, usize, usize),
+}
+
+impl fmt::Debug for PriorityEngine {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let fees = self
+            .priorities
+            .iter()
+            .enumerate()
+            .map(|(i, fees)| format!("{:>4} → {:?}", i, fees))
+            .join("\n");
+
+        write!(
+            f,
+            "[PriorityEngine] There is priority information for {} epochs:\n{}",
+            self.priorities.len(),
+            fees,
+        )
+    }
+}
+
+impl Default for PriorityEngine {
+    fn default() -> Self {
+        Self::with_capacity(DEFAULT_QUEUE_CAPACITY_EPOCHS)
+    }
+}
+
+impl<'de> Deserialize<'de> for PriorityEngine {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as Deserializer<'de>>::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Vec::<Priorities>::deserialize(deserializer).map(Self::from_vec)
+    }
+}
+
+impl Serialize for PriorityEngine {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        Serialize::serialize(&self.as_vec(), serializer)
+    }
+}
+
+/// Conveniently wraps a priority value with sub-nanoWit precision.
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+pub struct Priority {
+    nano_wit: u64,
+    sub_nano_wit: u64,
+}
```

Comment:
I can't recall exactly which one, but `f64` is missing `impl` for a trait that I was needing.

In addition to this, and on top of the implicit advantages of fixed point vs. floating point, using `f64` in this module would require extensive use of unnecessary `try_from`s here and there.

Also, personally, I like to give values with a semantic type that represents what they mean and provide convenience methods for handling those values nicely. As a matter of fact, only once I introduced this type I noticed that in some place I had used the fee instead of the priority. As both were using `u64`, the compiler was perfectly OK. In the same moment I tagged the priority as such, the compiler prevented me from making such a stupid fail.

Reply:
> I can't recall exactly which one, but f64 is missing impl for a trait that I was needing.

We use `OrderedFloat` to sort transactions according to priority in the transactions pool, so maybe that was the missing trait.